# Prime digit replacements
Source: https://projecteuler.net/problem=51

# Problem Statement
By replacing the 1st digit of the 2-digit number *3, it turns out that six of the nine possible values: 13, 23, 43, 53, 73, and 83, are all prime.

By replacing the 3rd and 4th digits of 56**3 with the same digit, this 5-digit number is the first example having seven primes among the ten generated numbers, yielding the family: 56003, 56113, 56333, 56443, 56663, 56773, and 56993. Consequently 56003, being the first member of this family, is the smallest prime with this property.

Find the smallest prime which, by replacing part of the number (not necessarily adjacent digits) with the same digit, is part of an eight prime value family.

### Planning
By some speculation I've broken down the problem to finding the smallest prime that, when added with multiples of 10110, yield a prime

In [8]:
def gen_next_prime(prev_primes, current_val):
    
    while True:
    
        is_prime = True

        for prime in prev_primes:
            if current_val%prime == 0:
                is_prime = False
                break
        
        if is_prime:
            break
        else:
            current_val += 1
    
    return current_val

In [9]:
def primes_to_n(n):
    """
    generates primes up to and one above n
    """
    
    primes = []
    
    if n > 0:
        primes.append(2)
    else:
        return primes
        
    while primes[-1] < n:
        next_prime = gen_next_prime(primes, primes[-1])
        
        primes.append(next_prime)
        
    return primes

In [13]:
## testing
primes_to_n(9)

[2, 3, 5, 7, 11]

In [16]:
import timeit

start = timeit.default_timer()

primes_to_1000000 = primes_to_n(10**6)

end = timeit.default_timer()
print(f"Time elapsed: {end-start}")

Time elapsed: 603.054835869702


In [17]:
'000'.count('0')

3

In [25]:
###
def n_zeros(int_list, n):
    """
    returns a list of strings in string_list that contain n zeros
    """
    
    string_list = [str(elem) for elem in int_list]
    
    n_zeros_list = []
    
    for elem in string_list:
        
        if elem.count('0') >= n:
            n_zeros_list.append( int(elem) )
            
    return  n_zeros_list

In [29]:
n_zeros_3 = n_zeros(primes_to_1000000, 3)
print(n_zeros_3)

[10007, 10009, 40009, 70001, 70003, 70009, 90001, 90007, 100003, 100019, 100043, 100049, 100057, 100069, 100103, 100109, 100207, 100403, 100501, 100609, 100703, 100801, 100907, 101009, 102001, 103001, 103007, 104003, 104009, 108007, 109001, 130003, 140009, 150001, 160001, 160009, 170003, 180001, 180007, 200003, 200009, 200017, 200023, 200029, 200033, 200041, 200063, 200087, 200201, 200401, 200407, 200609, 200807, 200903, 200909, 201007, 202001, 204007, 206009, 208001, 208003, 208009, 220009, 230003, 240007, 250007, 260003, 260009, 270001, 280001, 280009, 300007, 300017, 300023, 300043, 300073, 300089, 300109, 300301, 300809, 302009, 303007, 304009, 307009, 308003, 309007, 320009, 340007, 350003, 360007, 370003, 370009, 390001, 400009, 400031, 400033, 400051, 400067, 400069, 400087, 400093, 400109, 400207, 400307, 400409, 400601, 400607, 400703, 400903, 403001, 403003, 404009, 405001, 409007, 410009, 420001, 430007, 430009, 440009, 450001, 490001, 490003, 500009, 500029, 500041, 500057,

In [37]:
n_zeros_4 = n_zeros(primes_to_1000000, 4)
print(n_zeros_4)

[100003, 200003, 200009, 300007, 400009, 500009, 700001, 900001, 900007, 1000003]


In [39]:
n_zeros_2 = n_zeros(primes_to_1000000, 2)
print(n_zeros_2)

[1009, 2003, 3001, 4001, 4003, 4007, 5003, 5009, 6007, 7001, 8009, 9001, 9007, 10007, 10009, 10037, 10039, 10061, 10067, 10069, 10079, 10091, 10093, 10099, 10103, 10301, 10303, 10501, 10601, 10607, 10709, 10903, 10909, 11003, 12007, 13001, 13003, 13007, 13009, 14009, 16001, 16007, 19001, 19009, 20011, 20021, 20023, 20029, 20047, 20051, 20063, 20071, 20089, 20101, 20107, 20201, 20407, 20507, 20509, 20707, 20807, 20809, 20903, 21001, 22003, 23003, 24001, 24007, 26003, 28001, 29009, 30011, 30013, 30029, 30047, 30059, 30071, 30089, 30091, 30097, 30103, 30109, 30203, 30307, 30403, 30509, 30703, 30707, 30803, 30809, 32003, 32009, 36007, 37003, 40009, 40013, 40031, 40037, 40039, 40063, 40087, 40093, 40099, 40507, 40609, 40709, 40801, 40903, 43003, 45007, 49003, 49009, 50021, 50023, 50033, 50047, 50051, 50053, 50069, 50077, 50087, 50093, 50101, 50207, 50503, 50707, 50909, 51001, 52009, 53003, 54001, 55001, 55009, 56003, 56009, 59009, 60013, 60017, 60029, 60037, 60041, 60077, 60083, 60089, 6009

In [30]:
def replace_zeros(integer, digit):
    digit_str = str(digit)
    
    integer_str = str(integer)
    
    for idx in range(len(integer_str)):
        if integer_str[idx] == '0':
            integer_str = integer_str[0:idx] + digit_str + integer_str[idx+1:]
            
    return int(integer_str)

In [31]:
replace_zeros(1000, 4)

1444

In [43]:
def try_zero_primes(primes, zero_primes):
    """
    prime: set of primes with greatest value greater than greatest value in zero_primes
    zero_primes: list of primes that have a number of zero digits contained in them
    """
    
    for zero_prime in zero_primes:
        
        prime_count = 0
        
        for digit in [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]:
            if replace_zeros(zero_prime, digit) in primes:
                prime_count += 1
                
        if prime_count >= 7:
            return zero_prime
            
    return -1

In [44]:
## testing
try_zero_primes(set(primes_to_1000000), n_zeros_2)

56003

In [45]:
## testing
try_zero_primes(set(primes_to_1000000), n_zeros_3)

90007

In [46]:
## testing
try_zero_primes(set(primes_to_1000000), n_zeros_4)

-1

## Retry with slightly different conditions
Instead of only replacing 0 digits, try replacing 1s digits, since the former prevents the first digit from being changed. 

In [47]:
###
def n_ones(int_list, n):
    """
    returns a list of strings in string_list that contain n ones
    """
    
    string_list = [str(elem) for elem in int_list]
    
    n_zeros_list = []
    
    for elem in string_list:
        
        if elem.count('1') >= n:
            n_zeros_list.append( int(elem) )
            
    return  n_zeros_list

In [48]:
def replace_ones(integer, digit):
    digit_str = str(digit)
    
    integer_str = str(integer)
    
    for idx in range(len(integer_str)):
        if integer_str[idx] == '1':
            integer_str = integer_str[0:idx] + digit_str + integer_str[idx+1:]
            
    return int(integer_str)

In [59]:
def try_one_primes(primes, one_primes):
    """
    prime: set of primes with greatest value greater than greatest value in zero_primes
    zero_primes: list of primes that have a number of zero digits contained in them
    
    switches out the ones digits and tests if there are at least 8 digits for which the result is prime
    """
    
    for one_prime in one_primes:
        
        prime_count = 0
        
        for digit in [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]:
            if replace_ones(one_prime, digit) in primes:
                prime_count += 1
                
        if prime_count >= 8:
            return one_prime
            
    return -1

In [50]:
n_ones_3 = n_ones(primes_to_1000000, 3)
print(n_ones_3)

[1117, 1151, 1171, 1181, 1511, 1811, 2111, 4111, 8111, 10111, 10141, 10151, 10181, 10211, 10711, 11071, 11113, 11117, 11119, 11131, 11149, 11159, 11161, 11171, 11173, 11177, 11197, 11213, 11251, 11261, 11311, 11317, 11321, 11351, 11411, 11471, 11491, 11519, 11551, 11617, 11621, 11681, 11701, 11717, 11719, 11731, 11801, 11813, 11821, 11831, 11941, 11971, 11981, 12011, 12101, 12113, 12119, 12161, 12211, 12511, 12611, 12911, 13121, 13151, 13171, 13411, 13711, 14011, 14411, 15101, 15121, 15131, 15161, 15511, 16111, 16141, 16411, 16811, 17011, 17117, 17191, 17911, 18119, 18121, 18131, 18181, 18191, 18211, 18311, 18911, 19121, 19141, 19181, 19211, 21011, 21101, 21121, 21191, 21211, 21611, 21911, 22111, 25111, 26111, 28111, 31121, 31151, 31181, 31511, 35111, 40111, 41011, 41113, 41117, 41131, 41141, 41161, 41411, 41611, 41911, 44111, 47111, 50111, 51131, 51151, 51511, 58111, 61121, 61141, 61151, 61211, 61511, 65111, 68111, 70111, 71011, 71119, 71161, 71171, 71191, 71411, 71711, 79111, 80111, 

In [51]:
n_ones_4 = n_ones(primes_to_1000000, 4)
print(n_ones_4)

[10111, 11113, 11117, 11119, 11131, 11161, 11171, 11311, 11411, 16111, 101111, 101113, 101117, 101119, 101141, 101161, 101411, 101611, 109111, 110119, 110161, 110311, 110711, 111031, 111091, 111103, 111109, 111119, 111121, 111127, 111143, 111149, 111187, 111191, 111211, 111217, 111271, 111301, 111317, 111341, 111431, 111491, 111521, 111581, 111611, 111641, 111721, 111731, 111751, 111781, 111791, 111821, 111871, 111913, 111919, 112111, 112121, 112181, 113011, 113111, 113117, 113131, 113161, 113171, 114113, 114161, 114311, 115117, 115151, 115211, 115811, 116101, 116113, 116131, 116141, 116191, 116411, 116911, 117101, 117119, 117191, 117511, 117811, 117911, 118171, 118211, 118411, 119101, 119131, 119191, 119311, 119611, 121151, 121171, 121181, 121711, 128111, 131011, 131101, 131111, 131113, 131171, 131311, 131611, 131711, 136111, 140111, 141101, 141121, 141131, 141161, 141181, 141311, 141511, 141811, 142111, 143111, 149111, 151121, 151141, 151171, 152111, 154111, 161141, 161411, 161611, 1

In [60]:
### note: the zero digit must be disallowed or else numbers can be improperly constructed
try_one_primes(set(primes_to_1000000), n_ones_3)

111857

In [55]:
857 in primes_to_1000000

True

In [61]:
def try_one_primes(primes, one_primes):
    """
    prime: set of primes with greatest value greater than greatest value in zero_primes
    zero_primes: list of primes that have a number of zero digits contained in them
    
    switches out the ones digits and tests if there are at least 8 digits for which the result is prime
    """
    
    for one_prime in one_primes:
        
        prime_count = 0
        
        for digit in [1, 2, 3, 4, 5, 6, 7, 8, 9]: # 0 digit is disallowed
            if replace_ones(one_prime, digit) in primes:
                prime_count += 1
                
        if prime_count >= 8:
            return one_prime
            
    return -1

In [62]:
try_one_primes(set(primes_to_1000000), n_ones_3)

121313

# Solution:
121313 is the first prime whose digit-replacements can result in 8 different primes numbers

# Post-solution tinkering
Is it possible to generate primes faster than I did up above?

In [1]:
def gen_next_prime_fast(prev_primes, current_val):
    
    while True:
    
        is_prime = True

        for prime in prev_primes:
            
            ## additional code
            if prime > int(current_val**0.5) + 1: # note: without +1, an infinite loop on 3 is created
                break # no primes above this have a chance of being a factor
            if current_val%prime == 0:
                is_prime = False
                break

        if is_prime:
            break
        else:
            current_val += 1
    
    return current_val

In [5]:
# test
prev_primes = [2, 3, 5]
current_val = 6

print(gen_next_prime_fast(prev_primes, current_val))

7


In [16]:
def primes_to_n_fast(n):
    """
    generates primes up to and one above n
    """
    
    primes = []
    
    if n > 0:
        primes.append(2)
    else:
        return primes
        
    while primes[-1] < n:
        next_prime = gen_next_prime_fast(primes, primes[-1]+1)
        
        primes.append(next_prime)
        
        
    return primes

In [17]:
import timeit

start = timeit.default_timer()

primes_to_1000000_fast = primes_to_n_fast(1000000)

end = timeit.default_timer()
print(f"time taken: {end-start}")
# the time to beat is ~10 minutes



time taken: 9.351997887361449


This means that the total time of my solution is much faster than 10 minutes; it is hence viable by computational standards